### PR TITLE
speed up metadata dumping and loading for Redis

### LINKED ISSUE #670: Speed up metadata dump & load
**What would you like to be added**:
Speed up metadata dump & load

**Why is this needed**:
Better user experience

### PR DESCRIPTION
**Pull request by zhijian-pro**
Close #670

### CODE REVIEW

File: pkg/meta/redis.go [hunk 1]
```diff
@@ -2968,87 +2972,147 @@ func (r *redisMeta) checkServerConfig() {
...
 		entries[k] = entry
 		if showProgress != nil {
 			showProgress(0, 1)
 		}
 	}
-	return entries, nil
+	return entries
+}
+
+var snap *snapStruct
```
Comment (davies) [inline pkg/meta/redis.go:3029]: 放到 redisMeta 里吧

Comment (zhijian-pro) [reply pkg/meta/redis.go:3029]: ok

File: pkg/meta/redis.go [hunk 2]
```diff
@@ -2968,87 +2972,147 @@ func (r *redisMeta) checkServerConfig() {
...
-	st := m.txn(ctx, func(tx *redis.Tx) error {
-		a, err := tx.Get(ctx, m.inodeKey(inode)).Bytes()
-		if err != nil {
-			return err
-		}
-		attr := &Attr{}
-		m.parseAttr(a, attr)
-		e.Attr = dumpAttr(attr)
-		e.Attr.Inode = inode
+	a := []byte(snap.stringMap[m.inodeKey(inode)])
```
Comment (davies) [inline pkg/meta/redis.go:2977]: a could be nil

File: pkg/meta/redis.go [hunk 3]
```diff
@@ -2968,87 +2972,147 @@ func (r *redisMeta) checkServerConfig() {
...
+func (m *redisMeta) makeSnap() error {
+	snap = &snapStruct{
+		stringMap: make(map[string]string),
+		listMap:   make(map[string][]string),
+		hashMap:   make(map[string]map[string]string),
+	}
+
+	ctx := context.Background()
+	listType := func(keys []string) error {
+		for _, key := range keys {
```
Comment (davies) [inline pkg/meta/redis.go:3046]: use Pipeline to get the lists in a batch

File: pkg/meta/redis.go [hunk 4]
```diff
@@ -2968,87 +2972,147 @@ func (r *redisMeta) checkServerConfig() {
...
+		for i := 0; i < len(keys); i++ {
+			if s, ok := values[i].(string); ok {
+				snap.stringMap[keys[i]] = s
+			}
+		}
+		return nil
+	}
+
+	hashType := func(keys []string) error {
+		for _, key := range keys {
```
Comment (davies) [inline pkg/meta/redis.go:3070]: use pipeline to speed up this

File: pkg/meta/redis.go [hunk 5]
```diff
@@ -3067,22 +3131,19 @@ func (m *redisMeta) DumpMeta(w io.Writer) error {
 		length, _ := strconv.ParseUint(parts[1], 10, 64)
 		dels = append(dels, &DumpedDelFile{Ino(inode), length, int64(z.Score)})
 	}
-
-	tree, err := m.dumpEntry(m.root)
-	if err != nil {
-		return err
+	if err := m.makeSnap(); err != nil {
+		return errors.Errorf("Generate redis snapshot error: %v", err)
```
Comment (davies) [inline pkg/meta/redis.go:3135]: "fetch all metadata from Redis: %s"

File: pkg/meta/redis.go [hunk 6]
```diff
@@ -3283,11 +3341,25 @@ func (m *redisMeta) LoadMeta(r io.Reader) error {
...
+	maxNum := 10000
+	pool := make(chan struct{}, maxNum)
+	var g errgroup.Group
 	for _, entry := range entries {
-		if err = m.loadEntry(entry, counters, refs); err != nil {
-			return err
-		}
+		tmpE := entry
+		g.Go(func() error {
+			pool <- struct{}{}
```
Comment (davies) [inline pkg/meta/redis.go:3351]: move this out of goroutine

File: pkg/meta/redis.go [hunk 7]
```diff
@@ -3283,11 +3341,25 @@ func (m *redisMeta) LoadMeta(r io.Reader) error {
 
 	counters := &DumpedCounters{}
 	refs := make(map[string]int)
+
+	maxNum := 10000
```
Comment (davies) [inline pkg/meta/redis.go:3345]: 100

File: pkg/meta/redis.go [hunk 8]
```diff
@@ -3283,11 +3341,25 @@ func (m *redisMeta) LoadMeta(r io.Reader) error {
...
+	pool := make(chan struct{}, maxNum)
+	var g errgroup.Group
 	for _, entry := range entries {
-		if err = m.loadEntry(entry, counters, refs); err != nil {
-			return err
-		}
+		tmpE := entry
```
Comment (davies) [inline pkg/meta/redis.go:3349]: we should fail fast if err is not nil

**Pull request closed without merging**